File: src/leetcode/python/bit_manipulation/RepeatedDNASequences_187.py
class RepeatedDNASequences_187(object):
    # "[record, r][substring in record].add(substring)"?
    # if substring in record:
    #   r.add(substring)
    # else:
    #   record.add(substring)
    def findRepeatedDnaSequences(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        r, record = set(), set()
        for i in range(len(s) - 9):
            substring = s[i:i + 10]
            [record, r][substring in record].add(substring)
        return list(r)

File: src/leetcode/java/RepeatedDNASequences_187.java
package leetcode.java;

import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;

public class RepeatedDNASequences_187 {
  public List<String> findRepeatedDnaSequences(String s) {
    List<String> repeated = new LinkedList<>();
    Set<Integer> firstMatch = new HashSet<>();
    Set<Integer> secondMatch = new HashSet<>();
    char[] chmap = new char[26];
    chmap['A' - 'A'] = 0;
    chmap['C' - 'A'] = 1;
    chmap['G' - 'A'] = 2;
    chmap['T' - 'A'] = 3;
    
    for (int i = 0; i < s.length() - 9; i++) {
      int hash = 0;
      for (int j = i; j < i + 10; j++) {
        hash <<= 2;
        hash |= chmap[s.charAt(j) - 'A'];
      }
      if (!firstMatch.add(hash) && secondMatch.add(hash)) {
        repeated.add(s.substring(i, i + 10));
      }
    }
    return repeated;
  }
}
